Sources/IndiceNetworkClient/Models/APIError.swift

```diff
@@ -9,39 +9,37 @@ import Foundation
 
 public struct APIError: LocalizedError {
     
-    public struct SimpleError: NetworkError {
-        public let status: Int?
-        public let description: String
-    }
-    
-    public static let Unknown         = APIError(description: "Unknown Error")
-    public static let Unauthenticated = APIError(description: "Unauthenticated")
+    public static let Unknown         = APIError(description: "Unknown Error", code: nil)
+    public static let Unauthenticated = APIError(description: "Unauthenticated", code: nil)
+    public static let InvalidResponse = APIError(description: "Invalid HttpResponse from server", code: nil)
     
-    public let errorData: NetworkError
+
+    public let errorDescription: String
+    public let statusCode: Int?
     
-    public init(description: String, code: Int? = nil) {
-        self.init(errorData: SimpleError(status: code, description: description))
-    }
+    public let raw: Data?
     
-    public init(description: String) {
-        self.init(description: description, code: nil)
+    public init(description: String, code: Int?, data: Data? = nil) {
+        self.errorDescription = description
+        self.statusCode = code
+        self.raw = data
     }
     
-    public init(errorData: NetworkError) {
-        self.errorData = errorData
+    public init(response: HTTPURLResponse, data: Data?) {
+        self.init(description: response.description,
+                  code: response.statusCode,
+                  data: data)
     }
-    
-    public init<T: NetworkError>(response: HTTPURLResponse, withData data: Data?, type: T.Type) {
-        if let errorData = data {
-            if let problemDetails = try? JSONDecoder().decode(type, from: errorData) {
-                self.init(errorData: problemDetails)
-                return
-            }
-            if let errorDescription = String(data: errorData, encoding: .utf8) {
-                self.init(description: errorDescription, code: response.statusCode)
-                return
+
+    func getError<T: Decodable>(using decoder: DecoderProtocol?) -> T? {
```

Comment:
I love this. The ability to use a different decoder is future proofing the api client. I am wondering if you could potentially initialize a default decoder (JsonDecoder in your case) as a dependency from the parent class